### DISCUSSION TITLE
.toList() to .fromList()

### DISCUSSION THREAD

**Discussion by RamilNurtdinov** [nextflow-io/nextflow#3487] [Q&A] [answered]
### Expected behavior and actual behavior
Greetings from CRG. 
I want to run some channel and store the report in a (list) value 
I want to use this list many times.
From my point of view .toList() to .fromList() should be compatible.

### Steps to reproduce the problem

    current_list = Channel.from(1, 2, 3, 4).toList()
    current_list.view()
    executions = Channel.fromList( current_list )
    executions.view()

### Program output 

Cannot invoke method view() on null object

### Environment 

* Nextflow version: N E X T F L O W  ~  version 22.10.1
* Java version: openjdk 17.0.3-internal 2022-04-19
* Operating system: Linux
* Bash version: GNU bash, version 4.2.46(1)-release (x86_64-redhat-linux-gnu)
CRG cluster

**Comment by jemunro** [comment, marked as answer]
You can use the operator `flatMap` to perform the opposite action to `toList`
```nextflow
current_list = Channel.from(1, 2, 3, 4).toList()
current_list.view()
executions = current_list.flatMap()
executions.view()
```

**Reply by RamilNurtdinov** [reply]
Hi Jacob

This is exact solution I needed, thank you. 
I read my TSV index file and combine all lines inside with toList().
flatMap extracts necessary fields and makes new channel.

**Comment by mribeirodantas** [comment]
Hello, @RamilNurtdinov. You're under the misconception that `.toList()` returns a [regular Groovy] list, when actually it returns a Nextflow channel that consists of a sole element which is a list. The `fromList` channel factory expects a regular Groovy list, and that's why `fromList` and `toList` don't go hand-in-hand.

In DSL1, you could obtain the regular Groovy list from a channel outputted by `toList` through the `.value` property, but this does not work in DSL2 anymore, as it's anti-pattern. When writing your Nextflow pipeline, you should use the Nextflow Domain-Specific Language and, for Nextflow, information is always contained within a channel 😄

If you want to use a channel multiple times, this should be a value channel and that's exactly what the `toList` and `collect` operators do. If you want to iterate through the elements of this list within a channel, for example, you can use `subscribe`, `map`, among other operators.

**Reply by RamilNurtdinov** [reply]
Thank you Marcel

I got practical answer above that works in DSL2.
